### PR TITLE
RFE: add RISC-V 64-bit support

### PR DESCRIPTION
**Pull request by andreas-schwab**

### CODE REVIEW

**Comment by pcmoore** [comment]
@drakenclimber I'm thinking this is worth adding to the v2.5 milestone, assuming everything is working and look okay.  Thoughts?

**Comment by drakenclimber** [comment]
> @drakenclimber I'm thinking this is worth adding to the v2.5 milestone, assuming everything is working and look okay. Thoughts?

Yes, I agree.  This would be a really good addition for v2.5

**Comment by carlosedp** [comment]
Hi folks, any news on this? Is it scheduled to be included on next version?
Thanks!

**Comment by drakenclimber** [comment]
> Hi folks, any news on this? Is it scheduled to be included on next version?
> Thanks!

I'm still planning on trying to get it into v2.5.  (Not sure yet on v2.5's release date yet, but pressure is growing to get it out.)

I'll try to review this code this week.  Thanks!

File: include/seccomp.h.in [hunk 1]
```diff
@@ -196,6 +196,18 @@ struct scmp_arg_cmp {
...
 
+/**
+ * The RISC-V architecture tokens
+ */
+#define SCMP_ARCH_RISCV64	AUDIT_ARCH_RISCV64
+/* RISC-V support for audit was merged in 5.0-rc1 */
+#ifndef AUDIT_ARCH_RISCV64
+#ifndef EM_RISCV
+#define EM_RISCV		243
+#endif /* EM_RISCV64 */
```
Comment (drakenclimber) [inline include/seccomp.h.in]: nitpick - this close the `EM_RISCV` ifndef not `EM_RISCV64`

File: src/arch-riscv64-syscalls.c [hunk 2]
```diff
@@ -0,0 +1,539 @@
+#include <string.h>
```
Comment (drakenclimber) [inline src/arch-riscv64-syscalls.c]: Do we want some sort of copyright header here?  If nothing else perhaps the standard LGPL boilerplate?

File: src/arch-riscv64.c [hunk 3]
```diff
@@ -0,0 +1,17 @@
+#include <stdlib.h>
```
Comment (drakenclimber) [inline src/arch-riscv64.c]: Do we want some sort of copyright header here?  If nothing else perhaps the standard LGPL boilerplate?

File: src/arch-riscv64.h [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#ifndef _ARCH_RISCV64_H
```
Comment (drakenclimber) [inline src/arch-riscv64.h]: Do we want some sort of copyright header here?  If nothing else perhaps the standard LGPL boilerplate?

File: tools/util.h [hunk 5]
```diff
@@ -72,6 +72,13 @@
 #define AUDIT_ARCH_PPC64LE	(EM_PPC64|__AUDIT_ARCH_64BIT|__AUDIT_ARCH_LE)
 #endif
 
+#ifndef AUDIT_ARCH_RISCV64
+#ifndef EM_RISCV
+#define EM_RISCV		243
+#endif /* EM_RISCV64 */
```
Comment (drakenclimber) [inline tools/util.h]: same nitpick comment as above.

File: include/seccomp.h.in [hunk 6]
```diff
@@ -196,6 +196,18 @@ struct scmp_arg_cmp {
...
+/**
+ * The RISC-V architecture tokens
+ */
+#define SCMP_ARCH_RISCV64	AUDIT_ARCH_RISCV64
+/* RISC-V support for audit was merged in 5.0-rc1 */
+#ifndef AUDIT_ARCH_RISCV64
+#ifndef EM_RISCV
+#define EM_RISCV		243
+#endif /* EM_RISCV */
+#define AUDIT_ARCH_RISCV64	(EM_RISCV|__AUDIT_ARCH_64BIT|__AUDIT_ARCH_LE)
```
Comment (pcmoore) [inline include/seccomp.h.in]: In keeping with the other tokens in seccomp.h, I would have expected `AUDIT_ARCH_RISCV64` to be defined before it was used.  However, if that is the only nit in this review we can move that while merging.

**Comment by pcmoore** [comment]
Merged into master via 5432e15521d5ce5a7d3f26bf78674cbaa9d73d1f, thanks everyone!

@andreas-schwab please re-test with master just to make sure everything is still working (neither @drakenclimber or myself have RISC-V machines), and please make an attempt to test the libseccomp releases to make sure we don't break anything in the future :)

**Comment by carlosedp** [comment]
Amazing, thanks everyone!
Cc. @davidlt

**Comment by pcmoore** [comment]
@carlosedp if you've got a RISC-V system handy, we would appreciate some extra testing help :)

**Comment by andreas-schwab** [comment]
https://build.opensuse.org/package/live_build_log/home:Andreas_Schwab:Factory/libseccomp/r/riscv64

**Comment by carlosedp** [comment]
Hey @pcmoore, I've built and tested libseccomp on my RISC-V board. All good:

```
...
Regression Test Summary
 tests run: 5229
 tests skipped: 114
 tests passed: 5229
 tests failed: 0
 tests errored: 0
```

**Comment by pcmoore** [comment]
@andreas-schwab I see a single failure in your test run:
```
[ 5388s] Test 52-basic-load%%001-00001 result:   FAILURE 52-basic-load rc=22
```
... is this maybe due to the build/CI tool either not allowing seccomp policies to be loaded, or simply running a kernel that doesn't support seccomp-bpf?

**Comment by andreas-schwab** [comment]
What is needed to support seccomp-bpf?  The same error happens on armv7l, and that runs natively.  There are no restrictions on what a kernel can do in OBS.

**Comment by pcmoore** [comment]
At the very least the kernel needs to be build with support for `CONFIG_SECCOMP_FILTER`, which is dependent upon `HAVE_ARCH_SECCOMP_FILTER` that is defined in the arch/ABI specific kernel code.

Are you building your own kernels @andreas-schwab or are you using distro-kernels?

**Comment by andreas-schwab** [comment]
This is a normal distro build, as you can see from the logfile, and CONFIG_SECCOMP_FILTER is enabled in all kernels.

**Comment by pcmoore** [comment]
> This is a normal distro build, as you can see from the logfile, and CONFIG_SECCOMP_FILTER is enabled in all kernels.

Hmm, any chance you can help us debug why the load test was failing?

**Comment by pcmoore** [comment]
I'm going to reopen this issue until we can resolve this last problem.

Since @carlosedp got a clean run on a RISC-V board I'm guessing there is just some problem with the OBS build and not a larger problem with libseccomp on RISC-V.

**Comment by andreas-schwab** [comment]
The riscv64 failure is due to qemu-linux-user rejecting to emulate seccomp.

**Comment by andreas-schwab** [comment]
For the armv7l failure I see this:

prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0)  = 0
seccomp(SECCOMP_SET_MODE_STRICT, 1, NULL) = -1 EINVAL (Invalid argument)
seccomp(SECCOMP_SET_MODE_FILTER, 0, {len=4, filter=0x50e400}) = -1 EINVAL (Invalid argument)

**Comment by pcmoore** [comment]
Looks like I can't reopen this issue, bummer.

**Comment by pcmoore** [comment]
> The riscv64 failure is due to qemu-linux-user rejecting to emulate seccomp.

Okay, I think we can ignore that failure then, thank you.

> For the armv7l failure I see this:
```
prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0) = 0
seccomp(SECCOMP_SET_MODE_STRICT, 1, NULL) = -1 EINVAL (Invalid argument)
seccomp(SECCOMP_SET_MODE_FILTER, 0, {len=4, filter=0x50e400}) = -1 EINVAL (Invalid argument)
```

The fist call `prctl(PR_SET_NO_NEW_PRIVS, ...)` sets NO_NEW_PRIVS, which is expected (anc configurable via a filter attribute.  Regardless, it succeeds so there is nothing to worry about there.

The `seccomp(SECCOMP_SET_MODE_STRICT, 1, NULL)` is also expected as it is a quick test to see if the system supports the seccomp syscall, see the `sys_chk_seccomp_syscall()` function for more information.  In this case it appears to return EINVAL which indicates that the system does support the `seccomp()` syscall; this is good and expected.

However, the final syscall, `seccomp(SECCOMP_SET_MODE_FILTER, 0, {len=4, filter=0x50e400})` is a bit confusing as it returns EINVAL indicating an error.  The `len=4` tells us that the BPF filter contains four instructions, which I would expect (load ABI token, compare ABI token and jump to good/bad ABI, return ALLOW, return KILL).  Are you 100% that this kernel is build with `CONFIG_SECCOMP_FILTER`?

**Comment by andreas-schwab** [comment]
No, it doesn't, because it has OABI_COMPAT enabled.

**Comment by pcmoore** [comment]
Ah, ha!  Mystery solved :)

**Pull request closed without merging**